packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart

```diff
@@ -131,20 +142,104 @@ class FuchsiaDevice extends Device {
   @override
   Future<String> get sdkNameAndVersion async => 'Fuchsia';
 
-  _FuchsiaLogReader _logReader;
   @override
   DeviceLogReader getLogReader({ApplicationPackage app}) {
-    _logReader ??= _FuchsiaLogReader(this);
-    return _logReader;
+    return _logReader ??= _FuchsiaLogReader(this);
   }
+  _FuchsiaLogReader _logReader;
 
   @override
-  DevicePortForwarder get portForwarder => null;
+  DevicePortForwarder get portForwarder {
+    return _portForwarder ??= _FuchsiaPortForwarder(this);
+  }
+  _FuchsiaPortForwarder _portForwarder;
 
   @override
   void clearLogs() {
   }
 
   @override
   bool get supportsScreenshot => false;
+
+  /// Run `command` on the fuchsia device.
+  Future<String> run(String command) => fuchsiaSdk.run(this, command);
+
+  /// Finds the first port running a VM matching `isolateName` given `ports`.
+  ///
+  /// TODO(jonahwilliams): replacing this with the hub will require an update
+  /// to the flutter_runner.
+  Future<int> findIsolatePort(String isolateName, List<int> ports) => fuchsiaSdk.findIsolatePort(this, isolateName, ports);
+
+  /// List the ports currently running a dart observatory.
+  Future<List<int>> servicePorts() async {
+    final String lsOutput = await run('ls /tmp/dart.services');
+    return parseFuchsiaDartPortOutput(lsOutput);
+  }
+}
+
+class _FuchsiaPortForwarder extends DevicePortForwarder {
+  _FuchsiaPortForwarder(this.device);
+
+  final FuchsiaDevice device;
+
+  @override
+  Future<int> forward(int devicePort, {int hostPort}) async {
+    hostPort ??= 0;
+    // Note: the provided command works around a bug in -N, but the solution is flaky.
+    final List<String> command = <String>[
```

Comment:
In theory this should work, but it is > 95% flaky for me

Reply:
I think I saw an issue for this go by. Could you note the issue number here?